apiserver/apiserver.go

```diff
@@ -614,19 +605,52 @@ func (srv *Server) loop(ready chan struct{}) error {
 		}
 	}
 
+	controllerConfigService := srv.shared.controllerConfigService
+	controllerConfigWatcher, err := controllerConfigService.WatchControllerConfig()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := srv.catacomb.Add(controllerConfigWatcher); err != nil {
+		return errors.Trace(err)
+	}
+
 	close(ready)
+
 	srv.mu.Lock()
 	srv.healthStatus = "running"
 	srv.mu.Unlock()
 
-	<-srv.tomb.Dying()
+	ctx := srv.catacomb.Context(context.Background())
 
-	srv.mu.Lock()
-	srv.healthStatus = "stopping"
-	srv.mu.Unlock()
+	for {
+		select {
+		case <-srv.catacomb.Dying():
+			srv.mu.Lock()
+			srv.healthStatus = "stopping"
+			srv.mu.Unlock()
+
+			srv.wg.Wait() // wait for any outstanding requests to complete.
+			return errors.Trace(err)
```

Comment:
Since this will cause the runner to be killed with the same error, we can't return a `catacomb.ErrDying`. See https://github.com/juju/juju/pull/16187.